File: jar-common/src/main/java/com/beancounter/common/exception/RestApiException.java
package com.beancounter.common.exception;

import javax.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
@Slf4j
public class RestApiException {

  @ExceptionHandler({HttpMessageNotReadableException.class})
  public ResponseEntity<Object> handleBadRequest(HttpServletRequest request) {

    SpringExceptionMessage error = new SpringExceptionMessage();
    error.setError("We did not understand your request. Please reformat it and try again");
    error.setMessage("Message not readable");
    error.setPath(request.getRequestURI());
    error.setStatus(HttpStatus.BAD_REQUEST.value());
    return new ResponseEntity<>(error, new HttpHeaders(), HttpStatus.BAD_REQUEST);
  }

  @ExceptionHandler({DataIntegrityViolationException.class})
  public ResponseEntity<Object> handleIntegrity(HttpServletRequest request, Throwable e) {

    SpringExceptionMessage error = new SpringExceptionMessage();
    log.debug(e.getMessage());
    error.setError("Constraint violation in request. Work is not accepted");
    error.setMessage("Integrity Violation");
    error.setPath(request.getRequestURI());
    error.setStatus(HttpStatus.BAD_REQUEST.value());
    return new ResponseEntity<>(error, new HttpHeaders(), HttpStatus.BAD_REQUEST);
  }

}

File: jar-common/src/main/java/com/beancounter/common/exception/RestApiException.kt
package com.beancounter.common.exception

import org.slf4j.LoggerFactory
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import java.net.ConnectException
import java.util.*
import javax.servlet.http.HttpServletRequest

@ControllerAdvice
class RestApiException {
    companion object {
        private val log = LoggerFactory.getLogger(RestApiException::class.java)
    }

    @ExceptionHandler(ConnectException::class)
    fun handleSystemException(request: HttpServletRequest, e: Throwable): ResponseEntity<Any> {
        val error = SpringExceptionMessage(
                Date(),
                HttpStatus.INTERNAL_SERVER_ERROR.value(),
                "Unable to contact dependent system.",
                e.message, request.requestURI

        )
        log.error(e.message)
        return ResponseEntity(error, HttpHeaders(), HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(BusinessException::class)
    fun handleBusinessException(request: HttpServletRequest, e: Throwable): ResponseEntity<Any> {
        val error = SpringExceptionMessage(
                Date(),
                HttpStatus.BAD_REQUEST.value(),
                "We are unable to process your request.",
                e.message, request.requestURI

        )
        return ResponseEntity(error, HttpHeaders(), HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleBadRequest(request: HttpServletRequest): ResponseEntity<Any> {
        val error = SpringExceptionMessage(
                Date(),
                HttpStatus.BAD_REQUEST.value(),
                "We did not understand your request.",
                "Message not readable", request.requestURI)

        return ResponseEntity(error, HttpHeaders(), HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(DataIntegrityViolationException::class)
    fun handleIntegrity(request: HttpServletRequest, e: Throwable): ResponseEntity<Any> {
        val error = SpringExceptionMessage(
                Date(),
                HttpStatus.CONFLICT.value(),
                "Request rejected.",
                "Data integrity violation", request.requestURI)
        return ResponseEntity(error, HttpHeaders(), HttpStatus.BAD_REQUEST)
    }

}